### ISSUE TITLE
Downloading assets hangs on 24 out of 7...

### ISSUE THREAD

**Issue by lapsio** [CypherpunkArmory/UserLAnd#486]
Just when I thought hanging at 6 out of 7 is had...

![screenshot_20190101-082224](https://user-images.githubusercontent.com/1054079/50571004-ec2ee180-0d9e-11e9-88be-0868cc8979da.png)

I can't believe instability of files download has not been solved for like almost a year now... And that there's STILL no easy option to export logs of app itself... It's third time already when my proor becomes unbootable because app crashes due to issues with updates download :/

**Comment by lapsio** [comment]
After 7 bilions of app restarts it somehow managed to get up. Rly, there's no progress bar for single step, no warning that it can take half an hour to complete, zero verbosity, no log, nothing. You're stuck with spinning circle and being unable to figure out what is happening. It's frustrating af.

**Comment by MatthewTighe** [comment]
There will be a beta release today or tomorrow that will hopefully increase stability of downloads, as well as handling download failure. The Android framework component we've used to handle downloads in the past considers failed downloads as completed, and we hadn't previously done error checking on whether the download actually completed successfully, but we will now. 

We've also seen most instances of odd download numbers (like 24 out of 7) be a result of clicking more than once on a session or app before the progress bar popped up, and that won't be possible in the beta release either.

**Comment by lapsio** [comment]
Not really.

![screenshot_20190110-141006](https://user-images.githubusercontent.com/1054079/50972532-30c92400-14e7-11e9-9448-9b64ee9b11a8.png)

![screenshot_20190110-141006](https://user-images.githubusercontent.com/1054079/50972553-3aeb2280-14e7-11e9-9b02-77235ba879fe.png)

**Comment by lapsio** [comment]
At first I thought it may be caused by low space but I freed 6gb and still getting the same error

**Comment by MatthewTighe** [comment]
Is filesystem extraction failure repeatable for you? If so, a PRoot debug log leading to the illegal state would be useful.

**Comment by lapsio** [comment]
> Is filesystem extraction failure repeatable for you? If so, a PRoot debug log leading to the illegal state would be useful.

Yes it happens in 100% cases. I'm not sure if it *always* crashes at exactly the same point but it seems so. What verbosity level do you need? Heres level 9 log but it looks like severe case of overkill so idk if you just gonna filter it or should I collect another log.

https://lapsio.bestpony.ml/public/inet/PRoot_Debug_Log

**Comment by corbinlc** [comment]
That means the tar.gz file that provides the initial filesystem is corrupt or incomplete.  Looks like the failure happened earlier and was not caught.

**Comment by lapsio** [comment]
> That means the tar.gz file that provides the initial filesystem is corrupt or incomplete. Looks like the failure happened earlier and was not caught.

As I said I closed UserLAnd in the middle of assets downloading before last update (before download method change)

**Comment by corbinlc** [comment]
Even though we believe the new way of downloading files should be more robust, we are planning on adding checksum testing to find corrupt files.  We don't have this implemented yet.

**Comment by lapsio** [comment]
> Even though we believe the new way of downloading files should be more robust, we are planning on adding checksum testing to find corrupt files. We don't have this implemented yet.

I'm legit curious - are you planning to implement any form of filesystem export or support files cleaning? Because I know it wouldn't fix root of problems but it'd really simplify troubleshooting if anyone could simply wipe everything except filesystems and start from scratch.

**Comment by MatthewTighe** [comment]
An easier way is probably to offer a way for users to clean the support directly, as @lapsio suggested at one. That should allow redownload of corrupted files. If that didn't work it would mean either Android's `DownloadManager` component will _still_ show the download as actually successful and not just complete, or Kotlin's copy functions will complete successful when they should fail.

@lapsio Does this error occur again if you install a new filesystem of a type you have not downloaded before? We have planned to implement both, but they aren't explicitly on our roadmap. Exporting is a much more difficult problem, but I could whip up support file cleaning early next week.

**Comment by MatthewTighe** [comment]
@lapsio Have you had access to the filesystem in question at any point, i.e. been able to connect in a session? If so, it shouldn't be running extraction again anyway and that's a separate bug.

**Comment by lapsio** [comment]
> @lapsio Does this error occur again if you install a new filesystem of a type you have not downloaded before? We have planned to implement both, but they aren't explicitly on our roadmap. Exporting is a much more difficult problem, but I could whip up support file cleaning early next week.

I could install other distros at any time even before update. Here's full story:

1. I updated UserLAnd
2. It totally derped with downloading assets for my Debian filesystem (which is carried from the very first beta release of UserLAnd)
3. It became unbootable because each start attempt hanged at assets download like on first ss
4. Meanwhile I decided to try other distros, starting from Arch
5. But Arch again hanged at assets download so I got triggered and closed app
6. After bilion of retry attempts Debian finally managed to download assets and boot up.
7. Since it did I decided to retry Arch installation but it failed again
8. I tried other distro (Kali) and it worked just fine
9. I updated UserLAnd (from this point on we have new downloads mechanism)
10. I retried Arch installation and it still failed (with different error though, before it was failing at some .sh script)
11. it's reproducible

**Comment by MatthewTighe** [comment]
Hmm. It's a little worrisome that Debian continued to download assets. Do you remember how many it tried to download? Once a filesystem is successfully extracted it shouldn't ever download the tar.gz assets again, but without a more exact timeline it would be hard to guess at what other assets we may have updated at that time.

My guess for Arch is that the tar.gz files were just corrupted at some point. That should be manageable by implementing the "clean support files" feature. I'll make that my priority for the beginning of next week.

Your Debian install works successfully now though? It may download 1 asset when starting but shouldn't be downloading more.

**Comment by lapsio** [comment]
Debian downloads assets every once a while after UserLAnd updates. Number of assets being downloaded varies per update. Sometimes it doesn't download anything. Basically if it happens then it happens after app update. It's not full filesystem redownload. I'm not sure if it's even unpacking anything. Looks like downloading some additional scripts added by next release, idk.

**Comment by lapsio** [comment]
Download process is still incredibly unstable. I cleaned assets and tried to redownload. During that process I ran into billion of issues.

1. Files download process hangs (30+ min with zero network activity)

![screenshot_20190118-022901](https://user-images.githubusercontent.com/1054079/51364154-73888e80-1adb-11e9-873c-6b1e05412a76.png)

2. Sometimes counter derps

![screenshot_20190118-022400](https://user-images.githubusercontent.com/1054079/51364178-88fdb880-1adb-11e9-895d-d04998a85908.png)

3. Download process works completely randomly. It failed to download some parts of rootfs, yet installation process started. In the middle of installation I still saw Android download manager to still download some parts, then UserLAnd puked with hundred of errors like those:

![screenshot_20190118-042256](https://user-images.githubusercontent.com/1054079/51364331-f0b40380-1adb-11e9-9a00-5b745d8f91d6.png)

![screenshot_20190118-042657](https://user-images.githubusercontent.com/1054079/51364345-f7427b00-1adb-11e9-976e-749450969212.png)

Behavior of Android download manager is basically really random.

**Comment by lapsio** [comment]
Also I actually closed UserLAnd quite a while ago but Android downloads manager still keeps downloading random parts of rootfs.

**Comment by lapsio** [comment]
![screenshot_20190118-045404](https://user-images.githubusercontent.com/1054079/51364680-2efdf280-1add-11e9-85d4-e73e97eb6a58.png)

Here's list of files in Downloads directory, sorted by date

**Comment by MatthewTighe** [comment]
Have you tried the process without your VPN active? I've never used that service, and it's really hard to guarantee whether or not it could be interacting poorly with the Android download manager. 

UserLAnd listens to alerts from the Android download manager in its MainActivity. Do you put the app in the background while downloading things? That could potentially cause the issue. I'll look into how to improve the reliability of that.

**Comment by lapsio** [comment]
> Have you tried the process without your VPN active? I've never used that service, and it's really hard to guarantee whether or not it could be interacting poorly with the Android download manager.
> 
> UserLAnd listens to alerts from the Android download manager in its MainActivity. Do you put the app in the background while downloading things? That could potentially cause the issue. I'll look into how to improve the reliability of that.

Yes I typically I don't use VPN. I only used it in this case to show that there's no network activity (since VPN has built in traffic monitor I trust more than what system monitor shows)

Yes I noticed that moving UserLAnd to background completely breaks downloads tracking.

I also noticed that it's possible to glitch downloader and ttrick it to think that it downloaded everything when irl it didn't by starting download process several times, killing app several times and restarting it. UserLAnd seems to catch downloads initiated by previous attempts as own one increasing counter (which obviously at some point goes past 23/23 to something like 26/23 when app proceeds to launch even though in fact not all assets have been downloaded (I guess) just some of them have been counted twice. As result it again crashes with dozen of Illegal States.

I didn't really check nature of last behavior described. I can only tell that I managed to get past downloading process this way multiple times and that it always results in Illegal States and broken installation but I don't know exactly what causes counter to go up, double-counting downloads initiated by other attempts is just my educated guess.

**Comment by MatthewTighe** [comment]
Yeah, we check that we 'own' downloads coming in by checking the name of the download. That could probably made more robust by tracking download ids.

Hopefully no one is stress-testing our app as seriously in their average use, but I appreciate that you've found ways to break it that we haven't. I'll work on implemented some fixes.

**Comment by MatthewTighe** [comment]
FYI I don't think I'll have time to implement a robust enough solution for app closure during download process completely. I can make it so at least it will be a "clean" start every time downloads begin, but a solution to the larger problem I don't think is worth the effort right now. Additionally, it would probably negatively affect session startup time and I'm not sure that's worth it for this corner case either.

**Comment by lapsio** [comment]
I just tried to make it work... T-T

Stuck downloads are really frustrating since they realistically make app unusable. Actually breaking download process in way I described is for now the only way I can get to launch my old filesystem. Because download process never ends in "normal" way.

So ironically "fixing" detection of double-counted files would break UserLAnd for me even more since I couldn't complete download process at all :/

**Comment by lapsio** [comment]
I left phone yesterday for night to see what happens and it didn't notice downloaded assets for like 12h so I doubt It would ever do that.

I actually didn't manage to successfully download assets even once since I posted this Issue ticket. That's why I tried so much "weird" hacks to make it work. Without success.

**Comment by MatthewTighe** [comment]
I wish I could recreate this somehow. It really seems like the download manager is just failing to notify userland of completed downloads. I'll keep investigating our code to see if there's an error.

**Comment by lapsio** [comment]
Aaaand it's stuck again...
![screenshot_20190212-051249](https://user-images.githubusercontent.com/1054079/52611339-7c277700-2e85-11e9-8711-a3d9e42e8a59.png)

Now also closing app doesn't restart download process but "continues" however if app has been closed during download, even if that resumed download finishes just... nothing happens. And reoeated attemp to launch session restarts download from scratch.

BUT! This time I might just have found *some* correlation that may make some sense... possibly... of course it hasn't really been deeply tested but I noticed that download resumes every time I switch active app to Opera. Which also uses download manager. So _maybe_ there is some interference between some apps using download manager opened in background and UserLAnd. I finally managed to get those assets to download by quick switching between opera and UserLAnd.
